cli/src/cli_util.rs

```diff
@@ -3053,6 +3053,36 @@ fn handle_early_args(
     Ok(())
 }
 
+fn handle_shell_completion(
+    ui: &mut Ui,
+    app: &Command,
+    config: &config::Config,
+    cwd: &Path,
+) -> Result<(), CommandError> {
+    let mut string_args = vec![];
+    // Take the first two arguments as is, they must be passed to clap_complete
+    // without any changes. They are usually "jj --".
+    for arg_os in env::args_os().take(2) {
+        if let Some(string_arg) = arg_os.to_str() {
+            string_args.push(string_arg.to_owned());
+        } else {
+            return Err(cli_error("Non-utf8 argument"));
+        }
+    }
+    if env::args_os().nth(2).is_some() {
+        // Make sure aliases are expanded before passing them to
+        // clap_complete. We skip the first two args ("jj" and "--") for
+        // alias resolution, then we stitch the args back together, like
+        // clap_complete expects them.
+        let resolved_aliases = expand_args(ui, app, env::args_os().skip(2), config)?;
+        string_args.extend(resolved_aliases);
+    }
+    clap_complete::CompleteEnv::with_factory(|| app.clone())
+        .try_complete(string_args.iter(), Some(cwd))
+        .map(|_| ())
```

Comment:
nit: better to assert the return value to ensure that the `$COMPLETE` variable is respected?